src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_validators.py

```diff
@@ -4,22 +4,119 @@
+    for p in ns.key_ops or []:
+        if p not in allowed:
+            raise ValueError("unrecognized key operation '{}'".format(p))
+
+def validate_key_type(ns):
+    if ns.destination:
+        dest_to_type_map = {
+            'software': 'RSA',
+            'hsm': 'RSA-HSM'
+        }
+        ns.destination = dest_to_type_map[ns.destination]
+
+def validate_policy_permissions(ns):
+    key_perms = ns.key_permissions
+    secret_perms = ns.secret_permissions
+    cert_perms = ns.certificate_permissions
+
+    if not any([key_perms, secret_perms, cert_perms]):
+        raise argparse.ArgumentError(
+            None,
+            'specify at least one: --key-permissions, --secret-permissions, '
+            '--certificate-permissions')
+
+    key_allowed = [x.value for x in KeyPermissions]
+    secret_allowed = [x.value for x in SecretPermissions]
+    cert_allowed = [x.value for x in CertificatePermissions]
+
+    for p in key_perms or []:
+        if p not in key_allowed:
+            raise ValueError("unrecognized key permission '{}'".format(p))
+
+    for p in secret_perms or []:
+        if p not in secret_allowed:
+            raise ValueError("unrecognized secret permission '{}'".format(p))
+
+    for p in cert_perms or []:
+        if p not in cert_allowed:
+            raise ValueError("unrecognized cert permission '{}'".format(p))
+
+def validate_principal(ns):
     num_set = sum(1 for p in [ns.object_id, ns.spn, ns.upn] if p)
     if num_set == 0:
         raise argparse.ArgumentError(
-            None, 'One of the arguments --object-id --spn --upn is required.')
+            None, 'specify exactly one: --object-id, --spn, --upn')
     elif num_set > 1:
         raise argparse.ArgumentError(
-            None, 'Specify only --object-id, --spn or --upn.')
+            None, 'specify only one: --object-id, --spn, --upn')
 
-def process_set_policy_perms_namespace(ns):
-    if ns.perms_to_keys is None and ns.perms_to_secrets is None:
-        raise argparse.ArgumentError(
-            None, 'Specify at least --perms-to-keys or --perms-to-secrets.')
+def validate_resource_group_name(ns):
+    if not ns.resource_group_name:
+        vault_name = ns.vault_name
+        client = get_mgmt_service_client(KeyVaultManagementClient).vaults
+        for vault in client.list():
+            id_comps = parse_resource_id(vault.id)
+            if id_comps['name'] == vault_name:
+                ns.resource_group_name = id_comps['resource_group']
+                return
+        raise CLIError(
+            "The Resource 'Microsoft.KeyVault/vaults/{}'".format(vault_name) + \
+            " not found within subscription")
+
+# ARGUMENT TYPES
+
+def datetime_type(string):
+    ''' Validates UTC datettime in format '%Y-%m-%d\'T\'%H:%M\'Z\''. '''
+    date_format = '%Y-%m-%dT%H:%MZ'
+    return datetime.strptime(string, date_format)
 
+def vault_base_url_type(name):
+    return 'https://{}.vault.azure.net'.format(name)
```

Comment:
I've updated this to work with the other clouds in _azure_env.py. We will need to update this when the general multi-cloud solution is implemented (along with storage)